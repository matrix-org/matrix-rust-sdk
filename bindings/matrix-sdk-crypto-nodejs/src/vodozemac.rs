use napi_derive::*;

use crate::into_err;

/// An Ed25519 public key, used to verify digital signatures.
#[napi]
#[derive(Clone)]
pub struct Ed25519PublicKey {
    inner: vodozemac::Ed25519PublicKey,
}

#[napi]
impl Ed25519PublicKey {
    /// The number of bytes an Ed25519 public key has.
    #[napi(getter)]
    pub fn length(&self) -> u32 {
        vodozemac::Ed25519PublicKey::LENGTH as u32
    }

    /// Serialize an Ed25519 public key to an unpadded base64
    /// representation.
    #[napi]
    pub fn to_base64(&self) -> String {
        self.inner.to_base64()
    }
}

/// An Ed25519 digital signature, can be used to verify the
/// authenticity of a message.
#[napi]
pub struct Ed25519Signature {
    pub(crate) inner: vodozemac::Ed25519Signature,
}

impl From<vodozemac::Ed25519Signature> for Ed25519Signature {
    fn from(inner: vodozemac::Ed25519Signature) -> Self {
        Self { inner }
    }
}

#[napi]
impl Ed25519Signature {
    /// Try to create an Ed25519 signature from an unpadded base64
    /// representation.
    #[napi(constructor)]
    pub fn new(signature: String) -> napi::Result<Self> {
        Ok(Self {
            inner: vodozemac::Ed25519Signature::from_base64(signature.as_str())
                .map_err(into_err)?,
        })
    }

    /// Serialize a Ed25519 signature to an unpadded base64
    /// representation.
    #[napi]
    pub fn to_base64(&self) -> String {
        self.inner.to_base64()
    }
}

/// A Curve25519 public key.
#[napi]
#[derive(Clone)]
pub struct Curve25519PublicKey {
    inner: vodozemac::Curve25519PublicKey,
}

#[napi]
impl Curve25519PublicKey {
    /// The number of bytes a Curve25519 public key has.
    #[napi(getter)]
    pub fn length(&self) -> u32 {
        vodozemac::Curve25519PublicKey::LENGTH as u32
    }

    /// Serialize an Curve25519 public key to an unpadded base64
    /// representation.
    #[napi]
    pub fn to_base64(&self) -> String {
        self.inner.to_base64()
    }
}

/// Struct holding the two public identity keys of an account.
#[napi]
pub struct IdentityKeys {
    ed25519: Ed25519PublicKey,
    curve25519: Curve25519PublicKey,
}

#[napi]
impl IdentityKeys {
    /// The Ed25519 public key, used for signing.
    #[napi(getter)]
    pub fn ed25519(&self) -> Ed25519PublicKey {
        self.ed25519.clone()
    }

    /// The Curve25519 public key, used for establish shared secrets.
    #[napi(getter)]
    pub fn curve25519(&self) -> Curve25519PublicKey {
        self.curve25519.clone()
    }
}

impl From<matrix_sdk_crypto::olm::IdentityKeys> for IdentityKeys {
    fn from(value: matrix_sdk_crypto::olm::IdentityKeys) -> Self {
        Self {
            ed25519: Ed25519PublicKey { inner: value.ed25519 },
            curve25519: Curve25519PublicKey { inner: value.curve25519 },
        }
    }
}
