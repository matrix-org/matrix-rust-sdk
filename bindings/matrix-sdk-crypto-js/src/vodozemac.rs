//! Vodozemac types.

use wasm_bindgen::prelude::*;

/// An Ed25519 public key, used to verify digital signatures.
#[wasm_bindgen]
#[derive(Debug, Clone)]
pub struct Ed25519PublicKey {
    inner: vodozemac::Ed25519PublicKey,
}

#[wasm_bindgen]
impl Ed25519PublicKey {
    /// The number of bytes an Ed25519 public key has.
    #[wasm_bindgen(getter)]
    pub fn length(&self) -> usize {
        vodozemac::Ed25519PublicKey::LENGTH
    }

    /// Serialize an Ed25519 public key to an unpadded base64
    /// representation.
    #[wasm_bindgen(js_name = "toBase64")]
    pub fn to_base64(&self) -> String {
        self.inner.to_base64()
    }
}

impl From<vodozemac::Ed25519PublicKey> for Ed25519PublicKey {
    fn from(inner: vodozemac::Ed25519PublicKey) -> Self {
        Self { inner }
    }
}

/// An Ed25519 digital signature, can be used to verify the
/// authenticity of a message.
#[wasm_bindgen]
#[derive(Debug)]
pub struct Ed25519Signature {
    pub(crate) inner: vodozemac::Ed25519Signature,
}

impl From<vodozemac::Ed25519Signature> for Ed25519Signature {
    fn from(inner: vodozemac::Ed25519Signature) -> Self {
        Self { inner }
    }
}

#[wasm_bindgen]
impl Ed25519Signature {
    /// Try to create an Ed25519 signature from an unpadded base64
    /// representation.
    #[wasm_bindgen(constructor)]
    pub fn new(signature: String) -> Result<Ed25519Signature, JsError> {
        Ok(Self { inner: vodozemac::Ed25519Signature::from_base64(signature.as_str())? })
    }

    /// Serialize a Ed25519 signature to an unpadded base64
    /// representation.
    #[wasm_bindgen(js_name = "toBase64")]
    pub fn to_base64(&self) -> String {
        self.inner.to_base64()
    }
}

/// A Curve25519 public key.
#[wasm_bindgen]
#[derive(Debug, Clone)]
pub struct Curve25519PublicKey {
    inner: vodozemac::Curve25519PublicKey,
}

#[wasm_bindgen]
impl Curve25519PublicKey {
    /// The number of bytes a Curve25519 public key has.
    #[wasm_bindgen(getter)]
    pub fn length(&self) -> usize {
        vodozemac::Curve25519PublicKey::LENGTH
    }

    /// Serialize an Curve25519 public key to an unpadded base64
    /// representation.
    #[wasm_bindgen(js_name = "toBase64")]
    pub fn to_base64(&self) -> String {
        self.inner.to_base64()
    }
}

impl From<vodozemac::Curve25519PublicKey> for Curve25519PublicKey {
    fn from(inner: vodozemac::Curve25519PublicKey) -> Self {
        Self { inner }
    }
}

/// Struct holding the two public identity keys of an account.
#[wasm_bindgen(getter_with_clone)]
#[derive(Debug)]
pub struct IdentityKeys {
    /// The Ed25519 public key, used for signing.
    pub ed25519: Ed25519PublicKey,

    /// The Curve25519 public key, used for establish shared secrets.
    pub curve25519: Curve25519PublicKey,
}

impl From<matrix_sdk_crypto::olm::IdentityKeys> for IdentityKeys {
    fn from(value: matrix_sdk_crypto::olm::IdentityKeys) -> Self {
        Self {
            ed25519: Ed25519PublicKey { inner: value.ed25519 },
            curve25519: Curve25519PublicKey { inner: value.curve25519 },
        }
    }
}

/// An enum over the different key types a device can have.
///
/// Currently devices have a curve25519 and ed25519 keypair. The keys
/// transport format is a base64 encoded string, any unknown key type
/// will be left as such a string.
#[wasm_bindgen]
#[derive(Debug)]
pub struct DeviceKey {
    inner: matrix_sdk_crypto::types::DeviceKey,
}

impl From<matrix_sdk_crypto::types::DeviceKey> for DeviceKey {
    fn from(inner: matrix_sdk_crypto::types::DeviceKey) -> Self {
        Self { inner }
    }
}

#[wasm_bindgen]
impl DeviceKey {
    /// Get the name of the device key.
    #[wasm_bindgen(getter)]
    pub fn name(&self) -> DeviceKeyName {
        (&self.inner).into()
    }

    /// Get the value associated to the `Curve25519` device key name.
    #[wasm_bindgen(getter)]
    pub fn curve25519(&self) -> Option<Curve25519PublicKey> {
        use matrix_sdk_crypto::types::DeviceKey::*;

        match &self.inner {
            Curve25519(key) => Some((*key).into()),
            _ => None,
        }
    }

    /// Get the value associated to the `Ed25519` device key name.
    #[wasm_bindgen(getter)]
    pub fn ed25519(&self) -> Option<Ed25519PublicKey> {
        use matrix_sdk_crypto::types::DeviceKey::*;

        match &self.inner {
            Ed25519(key) => Some((*key).into()),
            _ => None,
        }
    }

    /// Get the value associated to the `Unknown` device key name.
    #[wasm_bindgen(getter)]
    pub fn unknown(&self) -> Option<String> {
        use matrix_sdk_crypto::types::DeviceKey::*;

        match &self.inner {
            Unknown(key) => Some(key.clone()),
            _ => None,
        }
    }

    /// Convert the `DeviceKey` into a base64 encoded string.
    #[wasm_bindgen(js_name = "toBase64")]
    pub fn to_base64(&self) -> String {
        self.inner.to_base64()
    }
}

impl From<&matrix_sdk_crypto::types::DeviceKey> for DeviceKeyName {
    fn from(device_key: &matrix_sdk_crypto::types::DeviceKey) -> Self {
        use matrix_sdk_crypto::types::DeviceKey::*;

        match device_key {
            Curve25519(_) => Self::Curve25519,
            Ed25519(_) => Self::Ed25519,
            Unknown(_) => Self::Unknown,
        }
    }
}

/// An enum over the different key types a device can have.
///
/// Currently devices have a curve25519 and ed25519 keypair. The keys
/// transport format is a base64 encoded string, any unknown key type
/// will be left as such a string.
#[wasm_bindgen]
#[derive(Debug)]
pub enum DeviceKeyName {
    /// The curve25519 device key.
    Curve25519,

    /// The ed25519 device key.
    Ed25519,

    /// An unknown device key.
    Unknown,
}
