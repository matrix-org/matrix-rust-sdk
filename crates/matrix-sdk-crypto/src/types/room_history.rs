/*
Copyright 2025 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//! Types for sharing encrypted room history, per [MSC4268].
//!
//! [MSC4268]: https://github.com/matrix-org/matrix-spec-proposals/pull/4268

use std::fmt::Debug;

use ruma::{DeviceKeyAlgorithm, OwnedRoomId};
use serde::{Deserialize, Serialize};
use vodozemac::{megolm::ExportedSessionKey, Curve25519PublicKey};

use crate::{
    olm::ExportedRoomKey,
    types::{
        deserialize_curve_key, events::room_key_withheld::RoomKeyWithheldContent,
        serialize_curve_key, EventEncryptionAlgorithm, SigningKeys,
    },
};
#[cfg(doc)]
use crate::{olm::InboundGroupSession, types::events::room_key::RoomKeyContent};

/// A bundle of historic room keys, for sharing encrypted room history, per
/// [MSC4268].
///
/// [MSC4268]: https://github.com/matrix-org/matrix-spec-proposals/pull/4268
#[derive(Deserialize, Serialize, Debug, Default)]
pub struct RoomKeyBundle {
    /// Keys that we are sharing with the recipient.
    pub room_keys: Vec<HistoricRoomKey>,

    /// Keys that we are *not* sharing with the recipient.
    pub withheld: Vec<RoomKeyWithheldContent>,
}

impl RoomKeyBundle {
    /// Returns true if there is nothing of value in this bundle.
    pub fn is_empty(&self) -> bool {
        self.room_keys.is_empty() && self.withheld.is_empty()
    }
}

/// An [`InboundGroupSession`] for sharing as part of a [`RoomKeyBundle`].
///
/// Note: unlike a room key received via an `m.room_key` message (i.e., a
/// [`RoomKeyContent`]), we have no direct proof that the original sender
/// actually created this session; rather, we have to take the word of
/// whoever sent us this key bundle.
#[derive(Deserialize, Serialize)]
pub struct HistoricRoomKey {
    /// The encryption algorithm that the session uses.
    pub algorithm: EventEncryptionAlgorithm,

    /// The room where the session is used.
    pub room_id: OwnedRoomId,

    /// The Curve25519 key of the device which initiated the session originally,
    /// according to the device that sent us this key.
    #[serde(deserialize_with = "deserialize_curve_key", serialize_with = "serialize_curve_key")]
    pub sender_key: Curve25519PublicKey,

    /// The ID of the session that the key is for.
    pub session_id: String,

    /// The key for the session.
    pub session_key: ExportedSessionKey,

    /// The Ed25519 key of the device which initiated the session originally,
    /// according to the device that sent us this key.
    #[serde(default)]
    pub sender_claimed_keys: SigningKeys<DeviceKeyAlgorithm>,
}

impl Debug for HistoricRoomKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("SharedRoomKey")
            .field("algorithm", &self.algorithm)
            .field("room_id", &self.room_id)
            .field("sender_key", &self.sender_key)
            .field("session_id", &self.session_id)
            .field("sender_claimed_keys", &self.sender_claimed_keys)
            .finish_non_exhaustive()
    }
}

impl From<ExportedRoomKey> for HistoricRoomKey {
    fn from(exported_room_key: ExportedRoomKey) -> Self {
        let ExportedRoomKey {
            algorithm,
            room_id,
            sender_key,
            session_id,
            session_key,
            sender_claimed_keys,
            shared_history: _,
            forwarding_curve25519_key_chain: _,
        } = exported_room_key;
        HistoricRoomKey {
            algorithm,
            room_id,
            sender_key,
            session_id,
            session_key,
            sender_claimed_keys,
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use ruma::{owned_room_id, DeviceKeyAlgorithm};
    use vodozemac::{
        megolm::ExportedSessionKey, Curve25519PublicKey, Curve25519SecretKey, Ed25519SecretKey,
    };

    use crate::types::{room_history::HistoricRoomKey, EventEncryptionAlgorithm};

    #[test]
    fn test_historic_room_key_debug() {
        let key = HistoricRoomKey {
            algorithm: EventEncryptionAlgorithm::MegolmV1AesSha2,
            room_id: owned_room_id!("!room:id"),
            sender_key: Curve25519PublicKey::from(&Curve25519SecretKey::from_slice(b"abcdabcdabcdabcdabcdabcdabcdabcd")),
            session_id: "id1234".to_owned(),
            session_key: ExportedSessionKey::from_base64("AQAAAAC2XHVzsMBKs4QCRElJ92CJKyGtknCSC8HY7cQ7UYwndMKLQAejXLh5UA0l6s736mgctcUMNvELScUWrObdflrHo+vth/gWreXOaCnaSxmyjjKErQwyIYTkUfqbHy40RJfEesLwnN23on9XAkch/iy8R2+Jz7B8zfG01f2Ow2SxPQFnAndcO1ZSD2GmXgedy6n4B20MWI1jGP2wiexOWbFS").unwrap(),
            sender_claimed_keys: vec![(DeviceKeyAlgorithm::Ed25519, Ed25519SecretKey::from_slice(b"abcdabcdabcdabcdabcdabcdabcdabcd").public_key().into())].into_iter().collect(),
        };

        assert_debug_snapshot!(key);
    }
}
