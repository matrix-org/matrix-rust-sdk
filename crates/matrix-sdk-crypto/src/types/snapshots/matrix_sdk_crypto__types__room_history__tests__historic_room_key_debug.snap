---
source: crates/matrix-sdk-crypto/src/types/room_history.rs
expression: key
---
SharedRoomKey {
    algorithm: "m.megolm.v1.aes-sha2",
    room_id: "!room:id",
    sender_key: "curve25519:2n4iKRAGOTwcw54MuivfjtU/KZl1Fslme/b1FZMMRQM",
    session_id: "id1234",
    sender_claimed_keys: SigningKeys(
        {
            "ed25519": Ed25519(
                "ed25519:kbNLK/pAgXV57gwDRd/0TjMp4BsqwtWXgPmSxgHVnpk",
            ),
        },
    ),
    ..
}
